library/src/cache/disk_io.rs

```diff
@@ -0,0 +1,104 @@
+use anyhow::{bail, Context};
+use serde::{de::DeserializeOwned, Serialize};
+use std::{
+    fs::File,
+    io::{BufReader, BufWriter},
+    path::Path,
+};
+
+// https://stackoverflow.com/questions/67087597/is-it-possible-to-use-rusts-log-info-for-tests
+#[cfg(test)]
+use std::println as debug; // Workaround to use println! for logs.
+
+pub fn write<S, P>(serializable: &S, path: &P) -> anyhow::Result<()>
+where
+    S: ?Sized + Serialize,
+    P: AsRef<Path>,
+{
+    debug!("Writing to {:?}", path.as_ref());
+
+    let path_as_ref = path.as_ref();
+    let containing_dir = path_as_ref
+        .parent()
+        .with_context(|| format!("Failed to get parent dir for {:?}", path_as_ref))?;
+
+    // Because File::create can sometimes fail if the full directory path doesn't exist,
+    // we create the directories in its path first.
+    std::fs::create_dir_all(containing_dir).with_context(|| {
+        format!(
+            "Failed to create root dir {:?} when initializing PatchManager",
+            path_as_ref
+        )
+    })?;
+
+    let file = File::create(path).context(format!("File::create for {:?}", path_as_ref))?;
+    let writer = BufWriter::new(file);
+    serde_json::to_writer_pretty(writer, serializable)
+        .context(format!("failed to serialize to {:?}", path_as_ref))
+}
+
+pub fn read<D, P>(path: &P) -> anyhow::Result<D>
+where
+    D: DeserializeOwned,
+    P: AsRef<Path>,
+{
+    debug!("Reading from {:?}", path.as_ref());
+
+    let path_as_ref = path.as_ref();
+    if !path_as_ref.exists() {
+        bail!("File {} does not exist", path_as_ref.display());
+    }
+
+    let file = File::open(path_as_ref)?;
+    let reader = BufReader::new(file);
+    serde_json::from_reader(reader)
+        .context(format!("failed to deserialize from {:?}", &path_as_ref))
```

Comment:
with_context will delay creation of the string if you think it matters.  context() will do it eagerly.

Reply:
I discovered `with_context` about halfway through writing this, thought I caught all of these. Fixed!